operator/internal/decommissioning/statefulset_decomissioner.go

```diff
@@ -0,0 +1,753 @@
+
+		s.delayedBrokerIDCache.Clean(setCacheKey)
+		// reqeueue if we have any pending volumes for decommission, and if not
+		// settle until the next thing trigger reconciliation
+		return s.delayedVolumeCache.Size(setCacheKey) > 0, nil
+	}
+
+	allNodes := collections.NewSet[int]()
+	allNodes.Add(health.AllNodes...)
+	downedNodes := collections.NewSet[int]()
+	downedNodes.Add(health.NodesDown...)
+	healthyNodes := allNodes.LeftDisjoint(downedNodes)
+
+	brokerOrdinalMap := map[int]collections.Set[int]{}
+	brokerMap := map[int]int{}
+	for _, brokerID := range health.AllNodes {
+		broker, err := adminClient.Broker(ctx, brokerID)
+		if err != nil {
+			log.Error(err, "fetching broker")
+			return false, err
+		}
+
+		// NB: We capture the ordinal here because it gives us the
+		// ability to sort and have an extra check to ensure that we're only decommissioning
+		// downed brokers who also have ordinals that are higher than what the stateful set
+		// in its current configuration would actually produce (i.e. we
+		// don't want to accidentally decommission any random unhealthy brokers).
+		//
+		// Additionally, any potential ordinal collisions tell us that maybe the node
+		// id has changed but the original pod is actually gone (i.e. like a ghost broker)
+		// when we have an ordinal collision, then we check to make sure we have at least one
+		// healthy node for the given ordinal before decommissioning.
+		ordinal, err := ordinalFromFQDN(broker.InternalRPCAddress)
+		if err != nil {
+			// continue since we can't tell whether we can decommission this or not
+			// but make a lot of noise about the fact that we can't map this back to
+			// an ordinal
+			log.Error(err, "unexpected error parsing broker pod ordinal", "address", broker.InternalRPCAddress, "broker", broker)
+			continue
+		}
+
+		if _, ok := brokerOrdinalMap[ordinal]; !ok {
+			brokerOrdinalMap[ordinal] = collections.NewSet[int]()
+		}
+
+		// NB: here we have potentially multiple brokers that align to the same internal RPC address
+		// if that's the case, then one of them is going to be bad and can be decommissioned
+		brokerOrdinalMap[ordinal].Add(brokerID)
+		brokerMap[brokerID] = ordinal
+	}
+
+	brokersToDecommission := []int{}
+	brokersToIgnore := []int{}
+	currentlyDecommissioningBrokers := []int{}
+
+	for _, downedNode := range health.NodesDown {
+		ordinal, ok := brokerMap[downedNode]
+		if !ok {
+			// skip because we can't actually determine whether we should
+			// decommission it or not without its ordinal
+			brokersToIgnore = append(brokersToIgnore, downedNode)
+		}
+
+		status, err := adminClient.DecommissionBrokerStatus(ctx, downedNode)
+		if err != nil {
+			if strings.Contains(err.Error(), "is not decommissioning") {
+				if ordinal >= requestedNodes {
+					// this broker is old and should be deleted
+					brokersToDecommission = append(brokersToDecommission, downedNode)
+					continue
+				}
+
+				brokers := brokerOrdinalMap[ordinal]
```

Comment:
At the line 490 the `ordinal` might be `0`, because it is the "pod-0" that was assigned Redpanda Node ID that needs to be decommissioned, but it can be `0` because `brokerMap` does not have entry for particular Redpanda Node ID. If it is the latter case this retrieval would panic, right?

Reply:
It shouldn't, the maps are constructed in lines 466-482, and should roughly look like this with something like `pod-0` with lets say 2 IDs of something random, like `8`, and `3` (assuming we have a ghost broker for `pod-0` and the broker that's current), in pseudo-code:

```
brokersMap = map[int]int{
  3: 0,
  8: 0,
}
brokersOrdinalMap = map[int]Set{
  0: Set(3, 8),
}
```

assuming broker with a node ID `3` is marked as unhealthy, line 490 is indexing the first map, such that:

```
nodeDown = 3
ordinal = 0
brokers = Set(3, 8)
```

and what the code below is doing is:

1. On line 507, if the `brokers` set just has 1 node, then ignore this since it's the current broker for the ordinal
2. On line 521 if any of the other brokers in the set (i.e. broker with node id `8`) are healthy, then we know they are the current broker for the ordinal, so we can decommission `downedNode` (i.e. `3`)
3. Otherwise (Line 539), we can't tell what the current broker is supposed to be, so don't decommission anything.